### PR TITLE
Dynamic manifest of model configs via storage URL

### PR DESCRIPTION
**Pull request by snormore**
The main purpose of this PR is to:
 - [x] Fetch a JSON manifest of models configs from a backend storage URL by specifying the `models_url` config. This is the same as adding to the existing deploy-time `models` config, but can be used as a dynamic manifest, since it expires the cached version every N minutes.

This PR also does this:
 - [x] Decouple and encapsulate the proxy layer code into the `Cluster` and new classes `Model`, `ModelService`, `Storage`, and `Manifest`.
 - [x] Support providing a `model_url` config as a less Docker-coupled config API, rather than just the `docker.environment.MODEL_URL`

TODO pre-merge:
 - [x] Backfill missing test coverage for public methods on `Model`, `ModelService`, `Storage`, and `Manifest`, you'll see the empty placeholder tests

### CODE REVIEW

File: tests/test_proxy_app_with_gs.py [hunk 1]
```diff
@@ -1,9 +1,33 @@
...
 
+@pytest.fixture
+def cluster_with_gs():
+    with tempfile.NamedTemporaryFile('w') as f:
+        f.write(json.dumps({
+            'models': {
+                'imagenet_mobilenet_v1': {
+                    'spec': 'mobilenet_v1',
+                    'replicas': 1,
+                    'docker': {
```
Comment (stephensolis) [inline tests/test_proxy_app_with_gs.py]: Is this right? Doesn't seem to match the new `cluster.yml`

Comment (snormore) [reply tests/test_proxy_app_with_gs.py]: There are 2 ways of defining the model URL, via this more Docker-coupled approach nested in the config and via the `model_url` at the top of the model config. Eventually it would be good to only support the latter, but this leaves open the ability to use custom base Docker images and define a replica count for now.
